core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java

```diff
@@ -0,0 +1,440 @@
+        {
+            // For all other nodes, don't let the context pass through.
+            List<PlanNode> children = node.getSources().stream()
+                    .map(child -> child.accept(this, new HashMap<>()))
+                    .collect(Collectors.toList());
+            return replaceChildren(node, children);
+        }
+
+        @Override
+        public PlanNode visitTableScan(TableScanNode node, Map<Symbol, SkewedContext> context)
+        {
+            final String qualifiedTableName = metadata.getTableMetadata(session, node.getTable()).getQualifiedName().toString();
+
+            if (this.skewedTables.containsKey(qualifiedTableName)) {
+                SkewJoinConfig skewConfig = this.skewedTables.get(qualifiedTableName);
+                Optional<Pair<Symbol, ColumnMetadata>> symbolAndMetadata = node.getAssignments().entrySet().stream()
+                        .map(entry -> new Pair<>(entry.getKey(), metadata.getColumnMetadata(session, node.getTable(), entry.getValue())))
+                        .filter(entry -> entry.getB().getName().equals(skewConfig.getColumnName()))
+                        .findFirst();
+                symbolAndMetadata.ifPresent(m -> {
+                    context.putIfAbsent(m.getA(), new SkewedContext(skewConfig.values, toSqlType(m.getB().getType())));
+                });
+            }
+
+            return node;
+        }
+    }
+
+    static class SkewJoinConfig
+    {
+        /*
+         * Parses and validates the user supplied session data.
+         * */
+
+        final String tableName;
+        final String columnName;
+        final List<String> values;
+
+        private static final int TABLE_INDEX = 0;
+        private static final int COLUMN_INDEX = 1;
+        private static final int VALUE_INDEX = 2;
+        private static final int MIN_SIZE = 3;
+
+        private SkewJoinConfig(List<String> rawMetadata)
+        {
+            tableName = rawMetadata.get(TABLE_INDEX);
+            columnName = rawMetadata.get(COLUMN_INDEX);
+            values = ImmutableList.copyOf(rawMetadata.subList(VALUE_INDEX, rawMetadata.size()));
+        }
+
+        public String getTableName()
+        {
+            return tableName;
+        }
+
+        public String getColumnName()
+        {
+            return columnName;
+        }
+
+        public List<String> getValues()
+        {
+            return values;
+        }
+
+        static Optional<SkewJoinConfig> getMetadata(List<String> rawMetadata)
+        {
+            if (rawMetadata.size() < MIN_SIZE) {
+                return Optional.empty();
+            }
+
+            return Optional.of(new SkewJoinConfig(rawMetadata));
+        }
+    }
+
+    static class JoinRewriter
+    {
+        private static final LongLiteral DEFAULT_PROJECT_VALUE = new LongLiteral("0");
+        private static final ArrayConstructor DEFAULT_UNNEST_VALUE = new ArrayConstructor(ImmutableList.of(DEFAULT_PROJECT_VALUE));
```

Comment:
We're trying to get rid of ArrayConstructor from the IR (see https://github.com/trinodb/trino/pull/14415/files). Replace this with an invocation of the internal `$array` function.